### ISSUE TITLE
Typescript support 

### ISSUE THREAD

**Issue by breath103** [mtraynham/lodash-joins#9]
Thanks for the great project first. 😄 
i'd like to convert this to typescript and send PR back to here, (which support js also obviously) are you ok with that? or should i just rewrite those on my own fork and keep it

**Comment by mtraynham** [comment]
Hi @breath103 , thanks for taking an interest in the library!  

I'd like to keep the code JavaScript for a few reasons; linting support, single source files, minimal webpack definition.  I am however, willing to add a TypeScript definition file and include it in the published package.

I started one, but my TypeScript is a bit rusty, so maybe you can advise me.  I didn't include the `lib/*.js` modules yet, but maybe that's something else that could be added.

**index.d.ts**
```ts
/// <reference types="lodash" />

interface IJoinAccessor<T, U> extends Function {
    (a: T): U
}

interface ICartesianProduct extends Function {
    (...arrays: any[][]): any[]
}

interface IJoin extends Function {
    <T, R>(left: T[], leftAccessor: IJoinAccessor<T, R>): T[];
    <T, U, R>(left: T[], leftAccessor: IJoinAccessor<T, R>, right: U[]): (T & U)[];
    <T, U, R>(left: T[], leftAccessor: IJoinAccessor<T, R>, right: U[], rightAccessor: IJoinAccessor<U, R>): (T & U)[];
}

interface INonMergeLeftJoin extends Function {
    <T, R>(left: T[], leftAccessor: IJoinAccessor<T, R>): T;
    <T, U, R>(left: T[], leftAccessor: IJoinAccessor<T, R>, right: U[]): T;
    <T, U, R>(left: T[], leftAccessor: IJoinAccessor<T, R>, right: U[], rightAccessor: IJoinAccessor<U, R>): T[];
}

interface INonMergeRightJoin extends Function {
    <T, R>(left: T[], leftAccessor: IJoinAccessor<T, R>): T[];
    <T, U, R>(left: T[], leftAccessor: IJoinAccessor<T, R>, right: U[]): U[];
    <T, U, R>(left: T[], leftAccessor: IJoinAccessor<T, R>, right: U[], rightAccessor: IJoinAccessor<U, R>): U[];
}

declare module 'lodash-joins' {
    import {LoDashStatic} from 'lodash';
    
    interface LoDashJoinsStatic extends LoDashStatic {
        cartesianProduct: ICartesianProduct;

        hashFullOuterJoin: IJoin;
        hashInnerJoin: IJoin;
        hashLeftAntiJoin: INonMergeLeftJoin;
        hashLeftOuterJoin: IJoin;
        hashLeftSemiJoin: INonMergeLeftJoin;
        hashRightAntiJoin: INonMergeRightJoin;
        hashRightOuterJoin: IJoin;
        hashRightSemiJoin: INonMergeRightJoin;

        nestedLoopFullOuterJoin: IJoin;
        nestedLoopInnerJoin: IJoin;
        nestedLoopLeftAntiJoin: INonMergeLeftJoin;
        nestedLoopLeftOuterJoin: IJoin;
        nestedLoopLeftSemiJoin: INonMergeLeftJoin;
        nestedLoopRightAntiJoin: INonMergeRightJoin;
        nestedLoopRightOuterJoin: IJoin;
        nestedLoopRightSemiJoin: INonMergeRightJoin;

        sortedMergeFullOuterJoin: IJoin;
        sortedMergeInnerJoin: IJoin;
        sortedMergeLeftAntiJoin: INonMergeLeftJoin;
        sortedMergeLeftOuterJoin: IJoin;
        sortedMergeLeftSemiJoin: INonMergeLeftJoin;
        sortedMergeRightAntiJoin: INonMergeRightJoin;
        sortedMergeRightOuterJoin: IJoin;
        sortedMergeRightSemiJoin: INonMergeRightJoin;
    }
    
    const lodashJoins: LoDashJoinsStatic;
    export default lodashJoins;
}
```

**Comment by breath103** [comment]
i'm actually surprised that it's almost identically same with the one that i wrote
```
function leftInnerJoin<T1, T2>(
  left: T1[], leftAccessor: (left: T1) => string,
  right: T2[], rightAccessor: (right: T1) => string) {
}
```
we might can talk about some Template argument namings, but since in most of case we'll use implicit typing it won't be a problem.
i think it seems pretty much good with this. do you have any plan to include this in NPM package in near future?

**Comment by mtraynham** [comment]
So I refactored it a bit here, https://github.com/mtraynham/lodash-joins/blob/master/lodash-joins.d.ts, and did some simple testing here, https://github.com/mtraynham/lodash-joins/blob/master/debug/typings.ts.  

Can you verify this works for you before I release and publish to NPM?

**Comment by breath103** [comment]
ok as of type definitions you provided, i confirmed that it's working all good 👍  good to go i think


beside, one thing i found it hard is that all of those methods returns "Merged" object. in other words **LeftType & RightType.**
now in typescript, this is totally fine if LeftType and rightType is interface.
such as `const x: LeftType & RightType`, x means any kind object (or class instance) that response to given interface. 

but if LeftType & RightType is "class instance", things gets complicated, and even breaking.
for example, 
```
class A {
  public id: number;
  public name: string;
  public nameLower() { return this.name.toLowerCase(); }
}

class B {
  public aId: number;
  public name: string;
  public nameLower() { return this.name.toLowerCase(); }
}

const a = new A();
a.name = "123"
a.id = 1

const b = new B();
b.aId = a.id
b.name = "321";

const c = Object.assign({}, a, b);

console.log(c.nameLower());
```

this totally passes Typescript compiler, but raise error at 
"console.log(c.nameLower());" since nameLower method is not exist in c object since it lost the prototype while Object.assign

so in short, how about returning { left: LeftType, right: RightType } for IJoin interfaces? which requires some changes in javascript for sure i guess

**Comment by mtraynham** [comment]
> so in short, how about returning { left: LeftType, right: RightType } for IJoin interfaces? which requires some changes in javascript for sure i guess

I value the idea, but I don't believe I want to head down that route.  Supporting a response that is different from the non-merge joins, such as anti & semi, seems a bit counter-intuitive.

Maybe instead, I could add support for a merger function on each of the merge joins (outer, inner, left/right outer).  It would take the two records that would currently be `Object.assign`ed, and instead give the user an option to do something else.  I might have it default to `Object.assign` for compatibility, but this would give you the option to return `{ left: LeftType, right: RightType }`.  

```ts
function defaultMerger (left: LeftType, right: RightType): LeftType & RightType {
     return Object.assign({}, left, right);
}

function leftRightMerger (left: LeftType, right: RightType): {left: LeftType, right: RightType} {
     return {left, right};
}
```

**Comment by mtraynham** [comment]
I've added a branch describing the solution above.  I believe this is probably sufficient for what you want to do, but can you verify this meets your use case.

https://github.com/mtraynham/lodash-joins/compare/merger_function

This is a pretty significant change so it deserves a major version bump.  I'll likely:
1. Release what is currently master as a patch version (2.0.3)
2. Merge the feature branch
3. Release master as a beta major version (3.0.0-beta1) in lieu of adding tests
4. Add tests
5. Release major version (3.0.0)

**Comment by breath103** [comment]
👍  this is much universal and backward compatible i can't wait to see this get merged!

**Comment by mtraynham** [comment]
@breath103 I've done 1-3 and I'm going to close this.  I'll need to add tests, but don't have the time at the moment.

npm has 2.0.3 as the lastest tag with the first TypeScript definition we discussed.  So if you run `npm install lodash-joins`, you'll get that version.

If you instead run `npm install lodash-joins@beta` or `npm install lodash-joins@3.0.0-beta.1`, you'll get the newer version we have discussed.

Thanks again for taking an interest in the library and let me know if you find a bug or have a feature.  Cheers 🍻 !

**Comment by breath103** [comment]
Great thanks! i'll use this on production right now so i'll let you know if i found any bugs haha

**Comment by FloNeu** [comment]
@breath103 @mtraynham Hey guys... sorry, i am totally lost right now - how would i import and use this in typescript?
Do i first import lodash using import * as _ from 'lodash'... and how to import lodash-joins to 'extend' lodash?

**Comment by mtraynham** [comment]
Hey @FloNeu - the code [re-exports Lodash with all the functionality appended to it](https://github.com/mtraynham/lodash-joins/blob/master/index.lodash.ts#L6-L32) as Lodash supports [mixins](https://lodash.com/docs/#mixin).  The [.d.ts](https://github.com/mtraynham/lodash-joins/blob/master/index.lodash.d.ts) file is setup the same way.

So you should be able to do the following:
```typescript
import _ from 'lodash-joins`;  // Lodash with join functions appended.

// General lodash functions
_.now()

// Join functions
_.hashFullOuterJoin(...)
```

Disregarding typescript for a moment, since I don't have a quick setup - a NodeJS REPL with CommonJS can show you a bit better what I mean when you dump the properties of `_`.
```javascript
var _ = require('lodash-joins');  // Import the default export, similar to "import _ from 'lodash-joins';

console.log(_);

<ref *1> [Function: lodash] {
  templateSettings: {
    escape: /<%-([\s\S]+?)%>/g,
    evaluate: /<%([\s\S]+?)%>/g,
    interpolate: /<%=([\s\S]+?)%>/g,
    variable: '',
    imports: { _: [Circular *1] }
  },
  after: [Function: after],
  ary: [Function: ary],
  assign: [Function (anonymous)],
  assignIn: [Function (anonymous)],
  assignInWith: [Function (anonymous)],
  assignWith: [Function (anonymous)],
  at: [Function (anonymous)],
  before: [Function: before],
  bind: [Function (anonymous)] { placeholder: [Circular *1] },
  bindAll: [Function (anonymous)],
  bindKey: [Function (anonymous)] { placeholder: [Circular *1] },
  castArray: [Function: castArray],
  chain: [Function: chain],
  chunk: [Function: chunk],
  compact: [Function: compact],
  concat: [Function: concat],
  cond: [Function: cond],
  conforms: [Function: conforms],
  constant: [Function: constant],
  countBy: [Function (anonymous)],
  create: [Function: create],
  // ... I removed alot of the Lodash functions from this example - there are many.
  VERSION: '4.17.21',
  cartesianProduct: [Function (anonymous)],
  hashFullOuterJoin: [Function (anonymous)],
  hashInnerJoin: [Function (anonymous)],
  hashLeftOuterJoin: [Function (anonymous)],
  hashLeftSemiJoin: [Function (anonymous)],
  hashLeftAntiJoin: [Function (anonymous)],
  hashRightOuterJoin: [Function (anonymous)],
  hashRightSemiJoin: [Function (anonymous)],
  hashRightAntiJoin: [Function (anonymous)],
  sortedMergeFullOuterJoin: [Function (anonymous)],
  sortedMergeInnerJoin: [Function (anonymous)],
  sortedMergeLeftOuterJoin: [Function (anonymous)],
  sortedMergeLeftSemiJoin: [Function (anonymous)],
  sortedMergeLeftAntiJoin: [Function (anonymous)],
  sortedMergeRightOuterJoin: [Function (anonymous)],
  sortedMergeRightSemiJoin: [Function (anonymous)],
  sortedMergeRightAntiJoin: [Function (anonymous)],
  nestedLoopFullOuterJoin: [Function (anonymous)],
  nestedLoopInnerJoin: [Function (anonymous)],
  nestedLoopLeftOuterJoin: [Function (anonymous)],
  nestedLoopLeftSemiJoin: [Function (anonymous)],
  nestedLoopLeftAntiJoin: [Function (anonymous)],
  nestedLoopRightOuterJoin: [Function (anonymous)],
  nestedLoopRightSemiJoin: [Function (anonymous)],
  nestedLoopRightAntiJoin: [Function (anonymous)]
}
```
